TCD1304.py

```diff
@@ -34,45 +32,56 @@
+        frequency = frequency_master_clock, #default frequency of masterclock 2 MHz; # no need to specify the variable type (float). View comment in L37
+        min_voltagerange_clock: float = 0, #it says that min_voltagerange_clock should (!) be a float. Otherwise it would not turn into a float automatically. Output on float will be 0.0 V and on int 0 V.
+        max_voltagerange_clock: float = 4,
+        pwmgen = ... # not sure if pwmgen needs to be assigned to a value ??
         ): 
 
-        self.min_masterclock_frequency = min_masterclock_frequency #Hier wird initialisiert, wenn mich nicht alles täsucht
+        self.min_masterclock_frequency = min_masterclock_frequency #initializing starts from here (? Not sure though)
         self.max_masterclock_frequency = max_masterclock_frequency
         self.frequency = frequency
-        self.min_volatagerange_clock = min_volatagerange_clock
-        self.max_volatagerange_clock = max_volatagerange_clock
-
+        self.min_voltagerange_clock = min_voltagerange_clock
+        self.max_voltagerange_clock = max_voltagerange_clock
+        self.pwmgen = PWMGenerator() #once assigned in __init__ self.pwmgen can be used in other methods
+        
     def power_source ():        # puts a voltage of 4V on PV1
         ps = PowerSupply()
         ps.pv1 = 4
         ps.pv1
     
-    def master_clock ():        # puts PWM with frequency of 2MHz on SQ1
-        pwmgen = PWMGenerator()
-        pwmgen.generate(["SQ1"], fmc, [0.5])       
+    def master_clock (self):        # puts PWM with frequency of 2MHz on SQ1; masterclock is the fastest clock
+        prescaler = int(math.log(oscillator_frequency / frequency_master_clock) / math.log(2))   # When setting a frequency by mapping the reference clock directly to a PWM output, only frequencies which are even factors of 128 MHz (the frequency of the PSLab's main oscillator) are available. The frequency is therefore not set by specifying the frequency itself, but by setting a prescaler.
+        self.pwmgen.map_reference_clock(["SQ1"], prescaler)    
     
-    def sh_clock ():            #  on SQ2 sets pulse for the integrationtime
+    def sh_clock (self):           #  on SQ2 sets pulse for the integrationtime
+        return self.pwmgen.generate() #is the return attribute the one to choose in order to set a second clock?
+
         ''' 
         integrationtime is started by the falling flank of voltageoutput. 
         the time for integration must be longer than 3_694 Elements of fmc.
         sh clock is started before icg is started. normal state of sh is a low
         level. high level can be quite short
         '''
 
-    def icg_clock ():           #  on SQ3 starts and stops the reading of the sensor
+    def icg_clock (self):           #  on SQ3 starts and stops the reading of the sensor; this third clock is the slowest clock
+        self.pwmgen.set_state()
+            for pulse in range(integration_elements):

```

Comment:
You need to define `integration_elements`. In https://tcd1304.files.wordpress.com/2016/07/tcd1304-timing-es.png it's 3694. I don't know if that's a constant value or if it can be something else, check the data sheet or the tcd1304 blog.

Reply:
Nevermind, I missed that you had already done so.